autoload/ale/path.vim

```diff
@@ -232,6 +238,12 @@ function! ale#path#ToURI(path) abort
 endfunction
 
 function! ale#path#FromURI(uri) abort
+    let l:uri_handler = ale#util#GetURIHandler(a:uri)
```

Comment:
Instead of checking for URI handlers in the function to convert a URI to a path you could check for a URI handler before turning a URI into a filename. This function is only supposed to convert `file://` into a path.

Reply:
See commit f397c2e.

This is something I was looking to address.

So there are two things in this commit.

1)
I realized that JDTToURI and URIToJDT are not needed. I thought that a vim buffer name had to conform to the OS's filename character set. I do not think this is the case. Thus there is no conversions needed to and from the buffer name.  textDocument/definition can retrieve a jdt:// URI, then it can be opened as a buffer (e.g., `:e jdt://...`). Then this jdt:// buffer name can be used directly as the URI for another textDocument/definition.

2)
To fix conflation of "path" and "URI", I renamed all calls to `ale#path#FromURI(uri)` to `ale#util#ToResource(uri)`, which is a wrapper like this:
```
function! ale#util#ToResource(uri) abort
    let l:uri_handler = ale#uri#GetURIHandler(a:uri)
    if l:uri_handler is# v:null
        " resource is a filesystem path
        let l:resource = ale#path#FromFileURI(a:uri)
    else
        " resource is a URI
        let l:resource = a:uri
    endif
    return l:resource
endfunction
```
The same is done with `ale#path#ToURI(path)` (wrapped in `ale#util#ToURI(resource)`). (Also `ale#util#ToURI` is renamed to `ale#util#ToFileURI`, and similarly, `ale#util#FromURI` is renamed to `ale#util#FromFileURI`. I renamed it temporarily but thought it might be more clear. I can change it back if you like.)

Basically, since now jdt:// URIs are used along with filesystem paths, the "path" is no longer inaccurate and so I generalized it to "resource". A resource can either be (1) a path or (2) a URI.

For example, this function,
```
function! ale#lsp#message#Definition(buffer, line, column) abort
    return [0, 'textDocument/definition', {
    \   'textDocument': {
    \       'uri': ale#path#ToURI(expand('#' . a:buffer . ':p')),
    \   },
    \   'position': {'line': a:line - 1, 'character': a:column - 1},
    \}]
endfunction
```
now looks like this:
```
function! ale#lsp#message#Definition(buffer, line, column) abort
    return [0, 'textDocument/definition', {
    \   'textDocument': {
    \       'uri': ale#util#ToURI(expand('#' . a:buffer . ':p')),
    \   },
    \   'position': {'line': a:line - 1, 'character': a:column - 1},
    \}]
endfunction
```
Its now `ale#util#ToURI(resource)` (the wrapper) instead of `ale#path#ToURI(path)`.

I also reorganized code. I put the `GetURLHandler` function in the file `autoload/ale/url.vim`. I also put jdt.vim in the directory `autoload/ale/url/`.

Let me know what you think of this possible solution.

Reply:
I think this generalization is useful as it would allow to handle other schemas without too much hassle.  I tested this on my machine and works very well.

Just one detail with the auto commands. There seems to be a problem with the load order of the files. In a particular case the autocommand defined in events.vim throws an error sayin  ale#util#ReadJDTLink() in unknown function.

To replicate the error:

 - Open a Java file.
 - Run :ALEFindReferences on a class name.
 - Select any of the jdt:// URIs in the references list.
 - See error happen.

On the other hand if we first call :ALEGoToDefinition this works and loads the jdt.vim file. After this calling ALEFindReferences works as expected sin the ale#util#ReadJDTLink() function is now defined.

Reply:
Thanks hsanson. 75e041d should fix that.